cmd/node-termination-handler.go

```diff
@@ -318,23 +318,22 @@ func drainOrCordonIfNecessary(interruptionEventStore *interruptioneventstore.Sto
 		runPreDrainTask(node, nodeName, drainEvent, metrics, recorder)
 	}
 
+        podNameList, err := node.FetchPodNameList(nodeName)
```

Comment:
I'm okay with your change to fix the pod list (thanks for that btw), but I'm not sure I agree that a webhook should be sent if `cordonNode/cordonAndDrainNode` return an err.  We weren't checking the err before it looks like (whoops), but I think it should check the err right after the `cordon*` if-branches and  if the err is not nil, then signal the interruptionEventStore.Workers  (not executing the PostDrainTask or webhook).  

I'm happy to make that change after this PR is merged since that seems to have been a latent bug.

Reply:
it looks like go-linter is failing and the Scheduled Maintenance Events dry-run test for IMDS mode e2e test is failing:

```
golangci-lint run --timeout=60m
cmd/node-termination-handler.go:327:3: ineffectual assignment to `err` (ineffassign)
		err = cordonNode(node, nodeName, drainEvent, metrics, recorder)
		^
cmd/node-termination-handler.go:329:3: ineffectual assignment to `err` (ineffassign)
		err = cordonAndDrainNode(node, nodeName, metrics, recorder, nthConfig.EnableSQSTerminationDraining)
		^
```

I think if you do what I mention above, it will fix this one. 

```
goroutine 51 [running]:
k8s.io/client-go/kubernetes/typed/core/v1.(*pods).List(0xc0000cefa0, 0x24fb2e8, 0xc0000ae018, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc00040af30, ...)
	/go/pkg/mod/k8s.io/client-go@v0.21.1/kubernetes/typed/core/v1/pod.go:95 +0xe7
github.com/aws/aws-node-termination-handler/pkg/node.Node.fetchAllPods(0x1, 0xc0000400ca, 0x18, 0xc00004a016, 0x46, 0x101, 0xc0000402b8, 0x9, 0xc0000461f8, 0x3, ...)
	/node-termination-handler/pkg/node/node.go:499 +0x175
github.com/aws/aws-node-termination-handler/pkg/node.Node.FetchPodNameList(0x1, 0xc0000400ca, 0x18, 0xc00004a016, 0x46, 0x101, 0xc0000402b8, 0x9, 0xc0000461f8, 0x3, ...)
	/node-termination-handler/pkg/node/node.go:366 +0x85
main.drainOrCordonIfNecessary(0xc000152540, 0xc000794000, 0x1, 0xc0000400ca, 0x18, 0xc00004a016, 0x46, 0x101, 0xc0000402b8, 0x9, ...)
	/node-termination-handler/cmd/node-termination-handler.go:321 +0x178
created by main.main
	/node-termination-handler/cmd/node-termination-handler.go:238 +0x12c8
```

I'm not entirely sure why this is happening. I'd have to dig into it more.

Reply:
IMHO webhook notification is important to notify of an interruption event even if NTH fails to cordon/drain the node for any reason. Additionally, NTH can send the reason for cordon/drain failure in webhook notification itself as another field in coming releases.

Reply:
It seems `fetchAllPods` tried to get the list of pods from a lost node. Not sure if I am entirely correct here for the below issue.
`panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x137c7a7]`

Reply:
I see your point on sending a webhook even if the drain fails. Ideally, that won't happen very often. I would like for the webhook to say that it failed though.

Reply:
`drainHelper.Client` was not initialised when dry-run was set to true and therefore it was referencing an invalid memory address. Since `fetchAllPods` requires `drainHelper.Client` to be initialised before calling the core v1 pod api. 
Now, all test cases passed after this change. Without this change all test cases with dry-run set to true were failing.

I will open another feature request for webhook changes to send the status of cordon/drain failures in the notification.

Reply:
Incase of `dry-run` as we are not initialising `drainHelper.Client` so `fetchAllPods` should also be not called.
This should fix the unit as well as e2e test cases. I tested this on my local for all e2e cases.

Reply:
👍  That looks like the right thing to do! :)   I'll take a closer look at the full code a little later, but I did pre-approve for tests to run. Thanks for doing this work and contributing to NTH!

Reply:
Hi @bwagner5,
If the changes look good to you, please ship it with v1.13.2.